Improve PIT context relocation

cbuescher opened the pull request:
This change improves point-in-time (PIT) context relocations in cases where we
loose a PIT due to e.g. node shutdowns or restarts.  We are already able to retry
shards where open PIT contexts have been lost in the case where re-create
identical reader contests on wither the same or a different node for read-only
and frozen indices (via FrozenEngine / ReadOnlyEngine). 
However, currently we are re-creating these contexts with every subsequent search 
because we don’t store the recreated context in a similar way as when we are opening 
a new PIT. This PR avoids this extra work by making the following changes: 

* map ReaderContexts held in SearchService by more than just an auto-incremented Long
id that can clash between different nodes.  Instead this PR changes the lookup to use the full 
ShardSearchContextId

* re-writing the PIT id we return with every search request in case where the results 
we receive show that the shards involved in answering the PIT request have been 
served by a different node than the one originally encoded in the PIT id

The later is possible because the in the original PIT design already provided
hooks for changing the PIT id between search request and we already document
that open point in time ids can change between requests. 
Currently, without re-writing the PIT ids, we can get e.g. into situations where 
we lose a shard PIT context due to node restart, but are able to re-create it on 
another node. If the shards in the group got reassigned to nodes different than 
the original node, subsequent requests can get routed for retries to alternating 
different nodes. This doesn’t matter in the current situation where we throw away each
re-created PIT context after use, but with this improvement that tracks these
context on the new nodes again, we want to make sure that subsequent searches go
to the same node.

cbuescher committed:
Improve PIT context relocation

This change improves point-in-time (PIT) context relocations in cases where we
loose a PIT due to e.g. node shutdowns or restarts. We are already able to retry
shards where open PIT contexts have been lost in the case where re-create
identical reader contests on wither the same or a different node for read-only
and frozen indices (via FrozenEngine / ReadOnlyEngine).
However, currently we are re-creating these contexts with every subsequent search
because we don’t store the recreated context in a similar way as when we are opening
a new PIT. This PR avoids this extra work.

cbuescher committed:
Update docs/changelog/135231.yaml

dnhatn commented:
@cbuescher Sure, I'll take a look today or tomorrow.

cbuescher committed:
Add keepalive to second and third query

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

dnhatn reviewed (commented):
Thanks, @cbuescher. I've left some questions and comments, but I think re-encoding the PIT when the target nodes have changed is a good idea.

cbuescher commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/RetrySearchIntegTests.java:
```diff
@@ -157,19 +178,45 @@ public void testRetryPointInTime() throws Exception {
+                    .setPointInTime(new PointInTimeBuilder(pitId).setKeepAlive(TimeValue.timeValueMinutes(2))),
+                resp -> {
+                    assertHitCount(resp, docCount);
+                    updatedPit.set(resp.pointInTimeId());
```
Good point. However, its not impossible the PIT id doesn't change. The test restarts all nodes. If all shards in the original PIT get allocated to their original nodes again we will reopen the context in the same location, leaving everything in that particular SearchContextIdForNode the same. When running tests I definitely saw both scenarios.

cbuescher commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -607,10 +614,70 @@ public void sendSearchResponse(SearchResponseSections internalSearchResponse, At
+        }
+    }
+
+    static <Result extends SearchPhaseResult> BytesReference maybeReEncodeNodeIds(
```
I assume you are referring to the functionality added with https://github.com/elastic/elasticsearch/pull/111516? If I understand that PR correctly we will have SearchContextIdForNode entries in the [PIT with a "null" node entry](https://github.com/elastic/elasticsearch/blob/73a174f00d2fd8239267dbf5dfa5e4b7f0d5a9b5/server/src/main/java/org/elasticsearch/action/search/SearchContextId.java#L104). I think in that case we won't [add that shard to the shard iterators](https://github.com/elastic/elasticsearch/blob/fb526d46a45a3f42c3df129502876ff8f35d6b6a/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L1942) of any subsequent search, so we won't get a Result for that shard. That is one reason why I added copying PIT id entries for everything that has no Result from the [old to the new encoded ID without change](https://github.com/elastic/elasticsearch/blob/73a174f00d2fd8239267dbf5dfa5e4b7f0d5a9b5/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java#L674).
Is that what you mean?

cbuescher commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -607,10 +614,70 @@ public void sendSearchResponse(SearchResponseSections internalSearchResponse, At
+                    updatedShardMap.put(shardId, original.shards().get(shardId));
+                }
+            }
+            return SearchContextId.encode(updatedShardMap, original.aliasFilter(), mintransportVersion, failures);
```
I just see TransportSearchAction goes to all shards, it would probably be something more selective to only the shards that had changes in location and use ClearScrollController#closeContexts directly.

cbuescher committed:
Adressing review comments

dnhatn commented:
@cbuescher Please let me know when it’s ready for another review. Thank you for working on it.

cbuescher committed:
Rework failure usage in updating PIT id

Commit:
[CI] Auto commit changes from spotless

cbuescher committed:
Adding feature flag

Only store retried contexts and rewrite PIT when the feature is enabled.

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -361,7 +363,8 @@ public class SearchService extends AbstractLifecycleComponent implements IndexEv
     private final AtomicLong idGenerator = new AtomicLong();
 
-    private final Map<Long, ReaderContext> activeReaders = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final Map<ShardSearchContextId, ReaderContext> activeReaders = ConcurrentCollections
```
Can we avoid this change? Hashing and comparing a ShardSearchContextId is more expensive than a Long.

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1264,25 +1264,45 @@ private ReaderContext findReaderContext(ShardSearchContextId id, TransportReques
+                if (PIT_RELOCATION_FEATURE_FLAG.isEnabled()) {
+                    // we are using a PIT here so set singleSession to false to prevent clearing after the search finishes
+                    readerContext = createAndPutReaderContext(
+                        contextId,
```
I think we should create a new contextId instead of reusing it?

dnhatn replied:
The sessionId of this contextId can be different from the sessionId of the SearchService.

dnhatn reviewed (commented):
@cbuescher Sorry for the delay—I missed your ping. I've left some comments, but the PR looks quite good now. Thank you!

cbuescher commented on server/src/main/java/module-info.java:
```diff
@@ -54,6 +54,7 @@
     requires org.apache.lucene.queryparser;
     requires org.apache.lucene.sandbox;
     requires org.apache.lucene.suggest;
+    requires jdk.jdi;
```
yep, most likely from some IDE action. Removed.

cbuescher committed:
Addressing review comments

cbuescher commented:
Hi @dnhatn, thanks a lot for the last review. I added a few of the smaller changes, left questions and comments for the others for your consideration before I go ahead and change anything else.

dnhatn commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -613,10 +624,93 @@ public void sendSearchResponse(SearchResponseSections internalSearchResponse, At
     }
 
     protected BytesReference buildSearchContextId(ShardSearchFailure[] failures) {
```
A shard-level request can fail for various reasons, such as a circuit breaking exception on a data node. In such cases, we try another node, and if successful, we replace the old context with the new one here. Alternatively, we could avoid replacing the search context, which means we do not need to close it. We can maintain a list of failure types (SearchContextMissingException or ...) that indicate the search contexts are no longer available and should be replaced. What do you think?

cbuescher replied:
Thanks for the explanation, I like the idea of conditioning the Id rewrite on failure types, however when looking at how to do this here in AbstractSearchAsyncAction I noticed that we clear earlier shard failures when [another shard responds with a result](https://github.com/elastic/elasticsearch/blob/613cf73ab373365de6cf718803d27bb3acc99dd9/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java#L523). I'm not sure yet how to make this possible here or if there are better locations in the code to do so. Need to do some digging.

cbuescher committed:
Fix AbstractSearchAsyncActionTests.testMaybeReEncode

cbuescher committed:
Add new ActiveReaders class encapsulating active readers and relocation mappings

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Minor change in node lookup

cbuescher committed:
Update PIT ids only when we saw SearchContextMissingException as reason to retry

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher commented:
Hi @dnhatn, thanks for the earlier direct feedback, I added [another change](https://github.com/elastic/elasticsearch/pull/135231#discussion_r2449669539) to the pit id update method, let me know if this helps. I think this is ready for another review then.

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
minor logging change

cbuescher committed:
Modify updating PIT id conditions

cbuescher committed:
Adressing review comments

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Correct condition for updating pit

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Going back to always close contexts if nodeId changed

This reverts commit bb0f2650af9c1c7ce5efbb8c9d3403abf021574b.

cbuescher committed:
Close context when concurrently retrying same id twice

cbuescher committed:
small cleanup

cbuescher committed:
more logger cleanups

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Separate normal and relocated putContext code paths

cbuescher committed:
minor cleanups

dnhatn commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1328,6 +1344,44 @@ final ReaderContext createAndPutReaderContext(
+                readerContext = null;
+                return finalReaderContext;
+            } else {
+                // we already have a mapping for this context, dont add a new one and use the existing instead
```
We might encounter a race condition between two retries: one tries to put the relocation key first, while the second detects that the relocation key already exists and tries to get the relocation reader. However, the first retry may not have added the reader to the map yet - only added the relocation key. This isn't a blocker, but it's something to be aware of.

dnhatn reviewed (commented):
Thanks @cbuescher. I've left two comments: one is optional for your awareness, and the other should be addressed before merging.

cbuescher committed:
Add RetrySearchIntegTests test for removed PIT context

This adds a test case to RetrySearchIntegTests that shows the current
ability to retry expired or closed PITs on read-only or frozen engines,
i.e. for searchable snapshots.

cbuescher commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1264,25 +1262,43 @@ private ReaderContext findReaderContext(ShardSearchContextId id, TransportReques
-                return createAndPutReaderContext(request, indexService, shard, searcherSupplier, defaultKeepAlive);
+                ReaderContext readerContext = null;
+                if (PIT_RELOCATION_FEATURE_FLAG.isEnabled()) {
+                    readerContext = createAndPutRelocatedPitContext(
```
This is an interesting aspect already of our current retry abilities for e.g. searchable snapshots I wasn't aware of. It seems we can already bring closed or expired PITs back to life if the underlying engine supports it (i.e. frozen/read-only). 
Let me try to rephrase your case to see if I understand the point correctly:
* we open a PIT shard context in session A which gets stored in the regular "activeReaders" map (i.e. id1 = `sessionID=A;id=10;readerId=abcde`)
* that context expires or gets closed for some reason, but the node stays up so the session is still A
* when we now retry with an "old" PIT id, we don't find the context and enter the relocation code path which with add a relocation mapping e.g. "id1 -> 123" and add the context in that position to the activeReaders

With the current logic we won't be able to remove the relocation mapping because removal with the current sessionId only goes to "activeReaders".
Is that reading correct so far?

I opened a test showing our current ability to retry even on closed/expired PITs for searchable snaphots at https://github.com/elastic/elasticsearch/pull/137371, it was surprising to me at first that this works but we probably need to keep it that way.
I considered several ideas to mitigate the problem but for the sake of simplicity would probably opt for excluding these kind of retries (same sessionId as current one) from the new relocation logic and improve on that later. I'll add a change along those lines in a bit.

cbuescher replied:
For simplicity, I opted for excluding contexts with similar session ids from the relocation logic for now in 1a302ac01f832d5e5a1901a09fb345748c64d3d and added the tests from #137371 to check we can still retry PITs on those cases, only without the improvements added here.

cbuescher committed:
Exemt context from current session from relocation handling

dnhatn commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -613,10 +625,91 @@ public void sendSearchResponse(SearchResponseSections internalSearchResponse, At
+        Logger logger
+    ) {
+        SearchContextId original = originalPit.getSearchContextId(namedWriteableRegistry);
+        boolean idChanged = false;
```
nit: I think we can replace this with `updatedShardMap != null` in line 693.

dnhatn commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -613,10 +625,91 @@ public void sendSearchResponse(SearchResponseSections internalSearchResponse, At
+            if (originalShard != null && originalShard.getSearchContextId() != null && originalShard.getSearchContextId().isRetryable()) {
+                // check if the node is different, if so we need to re-encode the PIT
+                String originalNode = originalShard.getNode();
+                if (originalNode != null && originalNode.equals(searchShardTarget.getNodeId()) == false) {
```
Can we compare the sessionId from `result.getContextId()` with the original sessionId instead?

cbuescher replied:
Unfortunately no. We need to keep the sessionId the same to the one encoded in the PIT up to this point. One major problem here is that we're [releasing irrelevant search contexts](https://github.com/elastic/elasticsearch/blob/fd2cc975418f16926bff08115c79d89c89c17114/server/src/main/java/org/elasticsearch/action/search/SearchPhase.java#L83) if e.g search phase result `ShardSearchContextId`s are not part of the requests PIT. I looked into this quite a bit and couldn't find an easy way around this. If really necessary, lets look into this in a follow up.

dnhatn commented on server/src/main/java/org/elasticsearch/search/ActiveReaders.java:
```diff
@@ -0,0 +1,91 @@
+    private final Map<ShardSearchContextId, Long> relocationMap = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final String sessionId;
+
+    ActiveReaders(String searchServiceSessionId, AtomicLong idGenerator) {
```
nit: idGenerator is not used.

dnhatn commented on server/src/main/java/org/elasticsearch/search/ActiveReaders.java:
```diff
@@ -0,0 +1,91 @@
+        assert previous == null;
+    }
+
+    Long generateRelocationMapping(ShardSearchContextId relocatedContextId, long mapppingKey) {
```
nit: spelling `mapppingKey`

dnhatn reviewed (approved):
I've left some nits and smaller comments, but LGTM. Thanks @cbuescher!

cbuescher committed:
Adressing review comments

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

cbuescher committed:
Adapt changelog

cbuescher committed:
Merge branch 'main' into improve-pit-relocation

dnhatn commented:
@cbuescher Thank YOU for all the iterations.
